crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -197,30 +219,82 @@ impl Observers {
                     entity,
                 },
                 data.into(),
+                &mut false,
             );
-        };
-
-        // Trigger observers listening for any kind of this trigger
-        observers.map.iter().for_each(&mut trigger_observer);
+        }
 
         // Trigger entity observers listening for this kind of trigger
         if entity != Entity::PLACEHOLDER {
-            if let Some(map) = observers.entity_observers.get(&entity) {
-                map.iter().for_each(&mut trigger_observer);
+            let mut entity = entity;
+            loop {
```

Comment:
Isn't it possible to get into an infinite loop if the traversal is cyclic? Do we need to add an upper bound to this? Is documentation on `Traversal` sufficient?

Reply:
It is totally possible to make an infinite loop. I've done some tweaking to the docs to make them clearer, and in general I am not expecting most users to implement Traversal themselves. Open to bike shedding.